### PR TITLE
feat: encrypt notify text data and add unit tests

### PR DESCRIPTION
**Pull request by sitaram-kalluri**
**- What I did**
1. Enhance notification text to support encryption/decryption of text messages.
2. Refactor NotificationServiceImpl for writing unit tests.
3. Introduce NotificationRequestTransformer class unit takes NotificationParams and returns a NotifyVerbBuilder.

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by sitaram-kalluri** [comment]
@gkc : The functional test and e2e tests are failing because the messageType is returned from the at_server. We are the PR with corresponding changes in at_server.

File: at_client/lib/src/response/at_notification.dart [hunk 1]
```diff
@@ -6,13 +8,16 @@ class AtNotification {
   late int epochMillis;
   String? value;
   String? operation;
+  String? messageType;
```
Comment (murali-shris) [inline at_client/lib/src/response/at_notification.dart:11]: can we have enum type here instead of String type

Comment (sitaram-kalluri) [reply at_client/lib/src/response/at_notification.dart:11]: Initially it was enum only. One problem with enum is we have to serialize when JSON encoding/decoding. Hence changed it to String.

**Comment by gkc** [comment]
Am I correct in believing that we need to hold on merging this to trunk until we've got the released the latest at_server and it's running across the entire production fleet?

File: at_client/lib/src/service/notification_service.dart [hunk 2]
```diff
@@ -156,10 +156,12 @@ class NotificationParams {
   }
 
   /// Returns [NotificationParams] to send a text message to another atSign.
-  static NotificationParams forText(String text, String whomToNotify) {
+  static NotificationParams forText(String text, String whomToNotify,
```
Comment (murali-shris) [inline at_client/lib/src/service/notification_service.dart:159]: can we replace whomToNotify with recipient ?

File: at_client/lib/src/service/notification_service_impl.dart [hunk 3]
```diff
@@ -180,43 +186,69 @@ class NotificationServiceImpl
...
+      var atKeyEncryption = AtKeyEncryptionManager.get(
+          notificationParams.atKey, _atClient.getCurrentAtSign()!);
+      // Get the NotifyVerbBuilder from NotificationParams
+      var builder = await NotificationRequestTransformer(
+              _atClient.getCurrentAtSign()!,
+              _atClient.getPreferences()!,
+              atKeyEncryption)
+          .transform(notificationParams);
+
+      if (builder.atKey!.startsWith(AT_PKAM_PRIVATE_KEY) ||
```
Comment (murali-shris) [inline at_client/lib/src/service/notification_service_impl.dart]: do we need this check? do we notify these pkam keys?

File: at_client/lib/src/service/notification_service.dart (hunk 2 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/service/notification_service.dart:159]: This might effect the backward compatibility of the app.

Comment (murali-shris) [reply at_client/lib/src/service/notification_service.dart:159]: got it.

File: at_client/lib/src/service/notification_service_impl.dart (hunk 3 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/service/notification_service_impl.dart]: No Murali, Will remove the IF condition and push the changes.

File: at_client/lib/src/response/at_notification.dart [hunk 4]
```diff
@@ -24,7 +29,9 @@ class AtNotification {
       'to': to,
       'epochMillis': epochMillis,
       'value': value,
-      'operation': operation
+      'operation': operation,
+      'messageType': messageType,
```
Comment (gkc) [inline at_client/lib/src/response/at_notification.dart:33]: What is 'messageType'?

File: at_client/lib/src/transformer/request_transformer/notify_request_transformer.dart [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import 'package:at_client/src/preference/at_client_preference.dart';
+import 'package:at_client/src/service/notification_service.dart';
+import 'package:at_client/src/transformer/at_transformer.dart';
+import 'package:at_client/src/util/at_client_util.dart';
+import 'package:at_commons/at_builders.dart';
+import 'package:at_commons/at_commons.dart';
+import 'package:at_client/src/encryption_service/encryption.dart';
+
+/// Class is responsible for taking the [NotificationParams] and converting into [NotifyVerbBuilder]
+class NotificationRequestTransformer
```
Comment (gkc) [inline at_client/lib/src/transformer/request_transformer/notify_request_transformer.dart:12]: I'd like to see some unit tests on this new class please.

File: at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart [hunk 6]
```diff
@@ -0,0 +1,49 @@
+import 'dart:async';
+
+import 'package:at_client/at_client.dart';
+import 'package:at_client/src/decryption_service/decryption.dart';
+import 'package:at_client/src/transformer/at_transformer.dart';
+import 'package:at_client/src/decryption_service/decryption_manager.dart';
+import 'package:meta/meta.dart';
+
+class NotificationResponseTransformer
```
Comment (gkc) [inline at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart:10]: I'd like to see some unit tests on this new class please.

File: at_client/lib/src/util/at_client_validation.dart [hunk 7]
```diff
@@ -120,4 +121,31 @@ class AtClientValidation {
       throw AtKeyException('namespace is mandatory');
     }
   }
+
+  Future<void> validateNotificationRequest(
```
Comment (gkc) [inline at_client/lib/src/util/at_client_validation.dart:125]: Do you have unit tests for this method?

File: at_client/lib/src/util/network_util.dart [hunk 8]
```diff
@@ -12,3 +12,9 @@ class NetworkUtil {
     return result;
   }
 }
+
+class NetworkConnectivityChecker {
```
Comment (gkc) [inline at_client/lib/src/util/network_util.dart]: Why is this here?

**Review by gkc — CHANGES_REQUESTED** [review]

**Comment by gkc** [comment]
@kalluriramkumar you wrote:
 > @gkc : The functional test and e2e tests are failing because the messageType is returned from the at_server. We are the PR with corresponding changes in at_server.

What have you done in order to make the functional and e2e tests pass?

File: at_client/lib/src/util/network_util.dart (hunk 8 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/util/network_util.dart]: I have removed this. I will double check on this.

File: at_client/lib/src/response/at_notification.dart (hunk 4 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/response/at_notification.dart:33]: The message is to indicate if the notification is of type Key or Text.

File: at_client/lib/src/transformer/request_transformer/notify_request_transformer.dart (hunk 5 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/transformer/request_transformer/notify_request_transformer.dart:12]: I have added unit tests in notification_service_test.dart file. May i know if the comments mean to add few more test cases?

File: at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart (hunk 6 above)
Comment (sitaram-kalluri) [reply at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart:10]: I have added unit tests in notification_service_test.dart file. May i know if the comments mean to add few more test cases?

Comment (gkc) [reply at_client/lib/src/transformer/response_transformer/notification_response_transformer.dart:10]: The unit tests are of the full notification; I think it would be great to have a few unit tests on **only** the behaviours of the Notification Request and Response Transformers

**Comment by sitaram-kalluri** [comment]
> @kalluriramkumar you wrote:
> 
> > @gkc : The functional test and e2e tests are failing because the messageType is returned from the at_server. We are the PR with corresponding changes in at_server.
> 
> What have you done in order to make the functional and e2e tests pass?

Apologies Gary, the functional tests and e2e tests have failed because of autoNotify made the ce2e atsigns inconsistent which i resolved. I missed correcting the my comment.

**Comment by gkc** [comment]
FYI server release v3.0.20 is currently being built for rollout today. Once rollout is complete and the outstanding comments on this PR have been addressed, then we can merge this PR and release new at_client version

**Review by murali-shris — APPROVED** [review]

File: at_end2end_test/test/notify_test.dart [hunk 9]
```diff
@@ -148,3 +149,50 @@ void main() {
     }
   });
 }
+
+/// Class responsible for getting the notifications from the cloud secondary
+class MonitorForNotification {
```
Comment (gkc) [inline at_end2end_test/test/notify_test.dart:154]: Nice

File: at_client/CHANGELOG.md [hunk 10]
```diff
@@ -1,3 +1,8 @@
+## 3.0.31
+- Enhance notify text to send text message encrypted
```
Comment (gkc) [inline at_client/CHANGELOG.md:2]: Please can you add to the changelog for the other PRs which have been merged since 3.0.30 before you publish this version 3.0.31

**Review by gkc — CHANGES_REQUESTED** [review]

File: at_client/CHANGELOG.md (hunk 10 above)
Comment (gkc) [reply at_client/CHANGELOG.md:2]: feat: Add to NotificationService.notify() signature: 
* added new optional callback parameter, onSentToSecondary
* added new optional 'checkForFinalDeliveryStatus' parameter
* added new optional 'checkForFinalDeliveryStatus' parameter
* and updated code documentation for NotificcationService.notify() method

Comment (sitaram-kalluri) [reply at_client/CHANGELOG.md:2]: Updated the CHANGELOG.md and pushed the changes.

**Review by gkc — APPROVED** [review]
